src/firefox/remote.js

```diff
@@ -86,27 +86,43 @@ export class RemoteFirefox {
     });
   }
 
+  getAddonsActor(): Promise<string> {
+    return new Promise((resolve, reject) => {
+      // getRoot should work since Firefox 55 (bug 1352157).
+      this.client.request('getRoot', (ignoredError, actors) => {
+        if (actors && actors.addonsActor) {
+          resolve(actors.addonsActor);
+          return;
+        }
+        log.debug('getRoot failed, falling back to listTabs');
+        // Fallback to listTabs otherwise, Firefox 49 - 77 (bug 1618691)
+        this.client.request('listTabs', (error, tabsResponse) => {
+          if (error) {
+            return reject(new WebExtError(
+              `Remote Firefox: listTabs() error: ${error}`));
+          }
+          if (!tabsResponse.addonsActor) {
+            log.debug(
+              'listTabs returned a falsey addonsActor: ' +
+              `${tabsResponse.addonsActor}`);
+            return reject(new RemoteTempInstallNotSupported(
+              'This is an older version of Firefox that does not provide an ' +
+              'add-ons actor for remote installation. Try Firefox 49 or ' +
+              'higher.'));
+          }
+          resolve(tabsResponse.addonsActor);
+        });
+      });
+    });
+  }
+
   installTemporaryAddon(
     addonPath: string
   ): Promise<FirefoxRDPResponseAddon> {
     return new Promise((resolve, reject) => {
-      this.client.request('listTabs', (error, tabsResponse) => {
-        if (error) {
-          return reject(new WebExtError(
-            `Remote Firefox: listTabs() error: ${error}`));
-        }
-        if (!tabsResponse.addonsActor) {
-          log.debug(
-            'listTabs returned a falsey addonsActor: ' +
-            `${tabsResponse.addonsActor}`);
-          return reject(new RemoteTempInstallNotSupported(
-            'This is an older version of Firefox that does not provide an ' +
-            'add-ons actor for remote installation. Try Firefox 49 or ' +
-            'higher.'));
-        }
-
+      this.getAddonsActor().then((addonsActor) => {
```

Comment:
Nit,  I see that we do reject explicitly if the `'installTemporaryAddon'` request rejects, but I'm wondering if the error we catch and reject there makes it clear if the request did fails because `addonsActor` was not defined.
Maybe `getAddonsActor` could still check that `addonsActor` it is going to resolve is actually defined and reject with a more clear error message if it is not (and so without the misleading error `This is an older version of Firefox...` in this case).

Reply:
> Nit, I see that we do reject explicitly if the `'installTemporaryAddon'` request rejects, but I'm wondering if the error we catch and reject there makes it clear if the request did fails because `addonsActor` was not defined.

Do you mean that it is not obvious from `installTemporaryAddon` that the error propagated through `.catch(reject)` is from `getAddonsActor`? Would this readability issue be resolved by using `.then(...., reject)` instead of `.then(....).catch(...)`?

> Maybe `getAddonsActor` could still check that `addonsActor` it is going to resolve is actually defined and reject with a more clear error message if it is not (and so without the misleading error `This is an older version of Firefox...` in this case).

`getAddonsActor` already checks whether `addonsActor` is defined before resolving. Could you clarify what you're requesting here?

Reply:
I see what you mean, but this comment wasn't about readability but about the fallback logic.

The current version does check if `getRoot` resolved but `addonsActor` is undefined, but at the moment it does fallback to `listTabs` in that case:

```
      this.client.request('getRoot', (ignoredError, actors) => {
        if (actors && actors.addonsActor) {
          resolve(actors.addonsActor);
          return;
        }
        ...
        // Fallback to listTabs otherwise, Firefox 49 - 77 (bug 1618691)
        this.client.request('listTabs', (error, tabsResponse) => {
          ...
```

if `getRoot` used to already include `addonsActor` when we introduced it for the first time in Firefox 55, then I think that we may instead reject if `ignoredError` and `actors.addonsActor` are both undefined, and fallback only if `getRoot` did explicitly fail.

How that sounds?